mediatek: filogic: add support for Tenbay WR3000K

jian-yu opened the pull request:
Tenbay WR3000K is an 802.11ax (Wi-Fi 6) router, based on MediaTek MT7981B. 

## Specifications:
- SoC: MetiaTek MT7981B
- RAM: Hynex H5TQ2G863GFR 512MiB
- Flash: Winbond W25N01GVZEIG 128MiB
- Wi-Fi: MediaTek MT7976C (2.4GHz/5GHz, 802.11ax, 2x2 MIMO, AX3000)
- MediaTek MT7915E: 2.4GHz and 5GHz
- Ethernet: 1x 10/100/1000 Mbps WAN + 3x 10/100/1000 Mbps LAN
- Switch: MediaTek MT7531AE
- UART: J4 (115200 baud)
- LEDs: Power
- Buttons: Reset, WPS
- PWR: 12V/1A DC, 5.5×2.1 connector

## MAC Addresses:
| Vendor  | OpenWrt Interface | Address       | Notes                                          |
|---------|-------------------|---------------|------------------------------------------------|
| WAN     | wan            | Stored MAC - 2     |  Label MAC |
| LAN     | br-lan             | Stored MAC - 1   |    |
| 2.4GHz  | phy0-ap0          | Stored MAC     |   Stored in  Factory + offset 4        |
| 5GHz    | phy1-ap0          | Stored MAC +1    |              |

## MTD Partitions:
- 0x000000000000-0x000000100000 : "BL2"
- 0x000000100000-0x000000180000 : "u-boot-env"
- 0x000000180000-0x000000380000 : "Factory"
- 0x000000380000-0x000000580000 : "FIP"
- 0x000000580000-0x000003580000 : "ubi"
- 0x000003580000-0x000006580000 : "ubi1"
- 0x000006580000-0x0000065a0000 : "Product"
- 0x0000065a0000-0x000007580000 : "Custom"

#### Note:
- The original partition-Ubi partition-Ubi1 is an AB dual system, and Openwrt only uses Ubi. So flash requires modifying the uboot variable `boot_from=ubi` to ensure that it only starts from Ubi.

- The `Product` and `Custom` partitions are original and only exist to align with the original layout; they are not used by OpenWrt.

## UBI Partitions (Dynamic):
- id: 0, kernel
- id: 1, rootfs
- id: 2, rootfs_data

## Flash instructions:
### Serial Flash
#### 1. Preparation

##### Hardware Requirements
- **USB-to-TTL Serial Adapter** (e.g., CH340 or CP2102).
- **Dupont Wires** (male-to-male, 3 wires).
- **PC/Laptop** with a serial communication tool.
- Screwdriver (to open the router case).

##### Software Requirements
1. **OpenWrt Firmware**:
   - Download the appropriate `wr3000k-<build_time>-mediatek-filogic-tenbay_wr3000k-squashfs-sysupgrade.bin` firmware file for your router from the [OpenWrt website](https://openwrt.org/).
2. **Serial Communication Tool**:
   - Windows: PuTTY, Tera Term.
   - Linux/Mac: Minicom, screen.
3. (Optional) **TFTP Server**:
   - Install a TFTP server like Tftpd64 or tftp-hpa.

---

#### 2. Setup Serial Connection

##### Hardware Connection
1. Open the router casing and locate the **TX, RX, and GND** pins.
2. Connect the router pins to the USB-to-TTL adapter as follows:
   - **TX (router)** → **RX (adapter)**
   - **RX (router)** → **TX (adapter)**
   - **GND (router)** → **GND (adapter)**
3. Do **not** connect the VCC pin to avoid damage.

##### Serial Tool Configuration
- **Baud rate**: 115200
- **Data bits**: 8
- **Stop bits**: 1
- **Parity**: None
- **Flow control**: None

---

#### 3. Modify U-Boot Boot_From Variable
1. Power on the router and observe the serial terminal output.
2. When prompted (e.g., `Hit any key to stop autoboot: 3`), press the '/' key quickly to interrupt the boot process.
3. You will see the U-Boot Boot Menu:
```plaintext
*** U-Boot Boot Menu ***

    1. Factory mode
    2. Startup system (Default)
    3. Upgrade firmware
    4. Upgrade ATF BL2
    5. Upgrade ATF FIP
    6. Upgrade single image
    7. Load image
    0. U-Boot console

Press UP/DOWN to move, ENTER to select, ESC/CTRL+C to quit
```
4. Select Option 0 by typing 0 and pressing Enter.
5. Input into
```plaintext
MT7981> setenv boot_from ubi
MT7981> saveenv
Saving Environment to MTD... Erasing on MTD device 'nmbm0'... OK
Writing to MTD device 'nmbm0'... OK
OK
MT7981> printenv
baudrate=115200
boot_from=ubi
...
```
the above indicates system will start from *ubi*.
and then type
```plaintext
MT7981> reset
```
will boot from *ubi*

#### 4. Enter U-Boot Upgrade Boot Menu

1. Power on the router and observe the serial terminal output.
2. When prompted (e.g., `Hit any key to stop autoboot: 3`), press the '/' key quickly to interrupt the boot process.
3. You will see the U-Boot Boot Menu:
```plaintext
*** U-Boot Boot Menu ***

    1. Factory mode
    2. Startup system (Default)
    3. Upgrade firmware
    4. Upgrade ATF BL2
    5. Upgrade ATF FIP
    6. Upgrade single image
    7. Load image
    0. U-Boot console

Press UP/DOWN to move, ENTER to select, ESC/CTRL+C to quit
```
4. Choose Option 3: Upgrade Firmware Enter Upgrade Mode
Select Option 3 by typing 3 and pressing Enter.
Upgrade Methods
You will be prompted to choose between:
```plaintext
*** Upgrading Firmware ***

Run image after upgrading? (Y/n): y

Available load methods:
    0 - TFTP client (Default)
    1 - Xmodem
    2 - Ymodem
    3 - Kermit
    4 - S-Record
    5 - RAM

Select (enter for default): 0

Input U-Boot's IP address: 192.168.1.1
Input TFTP server's IP address: 192.168.1.10
Input IP netmask: 255.255.255.0
Input file name: wr3000k-<build_time>-mediatek-filogic-tenbay_wr3000k-squashfs-sysupgrade.bin
```
Type Enter to proceed. The router will erase the old firmware and write the new one.

Signed-off-by: Jianyu Zhuang <xzjianyu@gmail.com>

ptpt52 commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,235 @@
+#include "mt7981.dtsi"
+
+/ {
+	model = "WR3000K";
```
model
`Tenbay WR3000K`

ptpt52 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1311,6 +1311,28 @@ define Device/ruijie_rg-x60-pro
+  UBINIZE_OPTS := -E 5
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  IMAGE_SIZE := 65536k
```
IMAGE_SIZE `49152k`

ptpt52 commented:
rebase and squash all commits to one

ptpt52 commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,235 @@
+	compatible = "tenbay,wr3000k", "mediatek,mt7981";
+
+	aliases {
+		ethernet0 = &gmac0;
```
should drop this line `ethernet0 = &gmac0`
see https://github.com/openwrt/openwrt/pull/17201

ptpt52 commented on package/boot/uboot-envtools/files/mediatek_filogic:
```diff
@@ -139,6 +139,9 @@ zyxel,ex5601-t0)
 	;;
+tenbay,wr3000k)
+        ubootenv_add_uci_config "/dev/mtd1" "0" "0x20000" "0x20000" 1
+        ;;
```
Here you need to place them in alphabetical order

jian-yu replied:
Corrected

jian-yu committed:
mediatek: filogic: add support for Tenbay WR3000K

Tenbay WR3000K is an 802.11ax (Wi-Fi 6) router, based on MediaTek MT7981B.

- SoC: MetiaTek MT7981B
- RAM: Hynex H5TQ2G863GFR 512MiB
- Flash: Winbond W25N01GVZEIG 128MiB
- Wi-Fi: MediaTek MT7976C (2.4GHz/5GHz, 802.11ax, 2x2 MIMO, AX3000)
- MediaTek MT7915E: 2.4GHz and 5GHz
- Ethernet: 1x 10/100/1000 Mbps WAN + 3x 10/100/1000 Mbps LAN
- Switch: MediaTek MT7531AE
- UART: J4 (115200 baud)
- LEDs: Power
- Buttons: Reset, WPS
- PWR: 12V/1A DC, 5.5×2.1 connector

| Vendor  | OpenWrt Interface | Address       | Notes                                          |
|---------|-------------------|---------------|------------------------------------------------|
| WAN     | wan            | Label MAC     |  Stored MAC in factory + offset 4, label MAC is Stored MAC - 2   |
| LAN     | br-lan             | Label MAC+1   |              |
| 2.4GHz  | phy0-ap0          | Label MAC + 2     |             |
| 5GHz    | phy1-ap0          | Label MAC + 3     |              |

- 0x000000000000-0x000000100000 : "BL2"
- 0x000000100000-0x000000180000 : "u-boot-env"
- 0x000000180000-0x000000380000 : "Factory"
- 0x000000380000-0x000000580000 : "FIP"
- 0x000000580000-0x000003580000 : "ubi"
- 0x000003580000-0x000006580000 : "ubi1"
- 0x000006580000-0x0000065a0000 : "Product"
- 0x0000065a0000-0x000007580000 : "Custom"

- The original partition-Ubi partition-Ubi1 is an AB dual system, and Openwrt only uses Ubi. So flash requires modifying the uboot variable `boot_from=ubi` to ensure that it only starts from Ubi.

- The Product and Custom partitions are original and only exist to align with the original layout; they are not used by OpenWrt.

- id: 0, kernel
- id: 1, rootfs
- id: 2, rootfs_data

- **USB-to-TTL Serial Adapter** (e.g., CH340 or CP2102).
- **Dupont Wires** (male-to-male, 3 wires).
- **PC/Laptop** with a serial communication tool.
- Screwdriver (to open the router case).

1. **OpenWrt Firmware**:
   - Download the appropriate `wr3000k-<build_time>-mediatek-filogic-tenbay_wr3000k-squashfs-sysupgrade.bin` firmware file for your router from the [OpenWrt website](https://openwrt.org/).
2. **Serial Communication Tool**:
   - Windows: PuTTY, Tera Term.
   - Linux/Mac: Minicom, screen.
3. (Optional) **TFTP Server**:
   - Install a TFTP server like Tftpd64 or tftp-hpa.

---

1. Open the router casing and locate the **TX, RX, and GND** pins.
2. Connect the router pins to the USB-to-TTL adapter as follows:
   - **TX (router)** → **RX (adapter)**
   - **RX (router)** → **TX (adapter)**
   - **GND (router)** → **GND (adapter)**
3. Do **not** connect the VCC pin to avoid damage.

- **Baud rate**: 115200
- **Data bits**: 8
- **Stop bits**: 1
- **Parity**: None
- **Flow control**: None

---

1. Power on the router and observe the serial terminal output.
2. When prompted (e.g., `Hit any key to stop autoboot: 3`), press the '/' key quickly to interrupt the boot process.
3. You will see the U-Boot Boot Menu:
```plaintext
*** U-Boot Boot Menu ***

    1. Factory mode
    2. Startup system (Default)
    3. Upgrade firmware
    4. Upgrade ATF BL2
    5. Upgrade ATF FIP
    6. Upgrade single image
    7. Load image
    0. U-Boot console

Press UP/DOWN to move, ENTER to select, ESC/CTRL+C to quit
```
4. Select Option 0 by typing 0 and pressing Enter.
5. Input into
```plaintext
MT7981> setenv boot_from ubi
MT7981> saveenv
Saving Environment to MTD... Erasing on MTD device 'nmbm0'... OK
Writing to MTD device 'nmbm0'... OK
OK
MT7981> printenv
baudrate=115200
boot_from=ubi
...
```
the above indicates system will start from *ubi*.
and then type
```plaintext
MT7981> reset
```
will boot from *ubi*

1. Power on the router and observe the serial terminal output.
2. When prompted (e.g., `Hit any key to stop autoboot: 3`), press the '/' key quickly to interrupt the boot process.
3. You will see the U-Boot Boot Menu:
```plaintext
*** U-Boot Boot Menu ***

    1. Factory mode
    2. Startup system (Default)
    3. Upgrade firmware
    4. Upgrade ATF BL2
    5. Upgrade ATF FIP
    6. Upgrade single image
    7. Load image
    0. U-Boot console

Press UP/DOWN to move, ENTER to select, ESC/CTRL+C to quit
```
4. Choose Option 3: Upgrade Firmware
Enter Upgrade Mode
Select Option 3 by typing 3 and pressing Enter.
Upgrade Methods
You will be prompted to choose between:
```plaintext
*** Upgrading Firmware ***

Run image after upgrading? (Y/n): y

Available load methods:
    0 - TFTP client (Default)
    1 - Xmodem
    2 - Ymodem
    3 - Kermit
    4 - S-Record
    5 - RAM

Select (enter for default): 0

Input U-Boot's IP address: 192.168.1.1
Input TFTP server's IP address: 192.168.1.10
Input IP netmask: 255.255.255.0
Input file name: wr3000k-<build_time>-mediatek-filogic-tenbay_wr3000k-squashfs-sysupgrade.bin
```
Type Enter to proceed. The router will erase the old firmware and write the new one.

Signed-off-by: Jianyu Zhuang <xzjianyu@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17172
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ptpt52 commented:
You can buy this device from Alibaba (https://www.alibaba.com/product-detail/Wi-Fi6-Router-WR3000K-2-4GHz_1600917316012.html). Hopefully it will be supported soon.

hauke commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+		compatible = "mediatek,eth-mac";
+		reg = <0>;
+		phy-mode = "2500base-x";
+		nvmem-cells = <&macaddr_factory_4 (-1)>;
```
The commit comment says the MAC is `Label MAC+1 ` this is using `-1`. Please fix the part which is wrong.

ptpt52 replied:
looks like that:
The commit message should be updated: `Stored MAC` is at `factory + offset 4` and the `label MAC` should be `MAC-2`

jian-yu replied:
Corrected

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1311,6 +1311,28 @@ define Device/ruijie_rg-x60-pro
+  DEVICE_MODEL := WR3000K
+  DEVICE_DTS := mt7981b-tenbay-wr3000k
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware uboot-envtools
```
```suggestion
  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware
```

`uboot-envtools` is already provided by the `DEFAULT_PACKAGES` in `target/linux/mediatek/filogic/target.mk`.

hauke commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+				read-only;
+			};
+
+			partition@100000 {
+				label = "u-boot-env";
+				reg = <0x0100000 0x0080000>;
+			};
```
Please add an entry to `package/boot/uboot-envtools/files/mediatek_filogic` too to make it possible to read and write the u-boot env from OpenWrt.

jian-yu commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1311,6 +1311,28 @@ define Device/ruijie_rg-x60-pro
+  DEVICE_DTS := mt7981b-tenbay-wr3000k
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware uboot-envtools
+  SUPPORTED_DEVICES := tenbay,wr3000k mt7981-spim-nand-gsw-wr3000k
```
`mt7981-spim-nand-gsw-wr3000k` also could be remove.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+/dts-v1/;
```
Missing SPDX-License?

aiamadeus commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+&eth {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mdio_pins>;
+
```
Extra empty lines.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+	switch0: switch@1f {
+		compatible = "mediatek,mt7531";
+		reg = <31>;
+		dsa,member = <0 0>;
```
Single switch does not need this.

jian-yu replied:
Corrected

aiamadeus commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+	pinctrl-0 = <&spi0_flash_pins>;
+	status = "okay";
+
+	spi_nand: flash@0 {
```
The 'spi_nand' is not needed.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+			partition@3580000 {
+			    label = "ubi1";
+			    reg = <0x3580000 0x3000000>;
+			 };
```
The ubi1 partition indent error.
The Product and Custom partitions are missing.

jian-yu replied:
Corrected

aiamadeus replied:
Still indent error.

aiamadeus commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+};
+
+&wifi {
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
Use this instead:
```
	nvmem-cells = <&eeprom_factory_0>;
	nvmem-cell-names = "eeprom";
````

jian-yu replied:
Corrected

aiamadeus commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1311,6 +1311,27 @@ define Device/ruijie_rg-x60-pro
+  PAGESIZE := 2048
+  IMAGE_SIZE := 49152k
+  KERNEL_IN_UBI := 1
+  IMAGES += factory.bin
```
Where is the factory.bin used?

jian-yu replied:
it used for uboot flash ubi(s) partitoin

jian-yu replied:
it used for uboot flash ubi(s) partition

aiamadeus replied:
Is sysupgrade.bin enough?

jian-yu replied:
factory.bin can use `dd` to flash mtd directly, so i think we can reserve it.

jian-yu commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+			};
+
+			factory: partition@180000 {
+				label = "Factory";
```
Corrected

jian-yu commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,234 @@
+
+			partition@380000 {
+				label = "FIP";
+				reg = <0x380000 0x0200000>;
```
Corrected

hauke commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,245 @@
+	};
+};
+
+&wifi {
+	nvmem-cells = <&eeprom_factory_0>;
+	nvmem-cell-names = "eeprom";
+	status = "okay";
+};
```
This does not build:
```
====== Make errors from logs/target/linux/dtb.txt ======
/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-mediatek_filogic/image-mt7981b-openwrt-one.dtb: Warning (i2c_bus_reg): Failed prerequisite 'reg_format'
/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-mediatek_filogic/image-mt7981b-openwrt-one.dtb: Warning (spi_bus_reg): Failed prerequisite 'reg_format'
../dts/mt7981b-openwrt-one.dts:445.9-449.4: Warning (avoid_default_addr_size): /soc/wifi@18000000/band@0: Relying on default #address-cells value
../dts/mt7981b-openwrt-one.dts:445.9-449.4: Warning (avoid_default_addr_size): /soc/wifi@18000000/band@0: Relying on default #size-cells value
../dts/mt7981b-openwrt-one.dts:451.9-455.4: Warning (avoid_default_addr_size): /soc/wifi@18000000/band@1: Relying on default #address-cells value
../dts/mt7981b-openwrt-one.dts:451.9-455.4: Warning (avoid_default_addr_size): /soc/wifi@18000000/band@1: Relying on default #size-cells value
/__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-mediatek_filogic/linux-6.6.65/arch/arm64/boot/dts/mediatek/mt7981.dtsi:715.23-733.5: ERROR (phandle_references): /soc/wifi@18000000: Reference to non-existent node or label "eeprom_factory_0"

  also defined at ../dts/mt7981b-tenbay-wr3000k.dts:241.7-245.3
ERROR: Input tree has errors, aborting (use -f to force output)
make[4]: *** [Makefile:43: /__w/openwrt/openwrt/openwrt/build_dir/target-aarch64-openwrt-linux-musl_musl/linux-mediatek_filogic/image-mt7981b-tenbay-wr3000k.dtb] Error 2
make[3]: *** [Makefile:18: dtb] Error 2
make[2]: *** [Makefile:12: dtb] Error 2
time: target/linux/dtb#1.96#1.10#3.02
```

ptpt52 commented on target/linux/mediatek/dts/mt7981b-tenbay-wr3000k.dts:
```diff
@@ -0,0 +1,249 @@
+			};
+
+			partition@3580000 {
+			    label = "ubi1";
+			    reg = <0x3580000 0x3000000>;
```
Thanks for merging this.
There is still a space indent issue here, please help fix it. @hauke

rtrabold commented:
Still not in 24.10 rc6? :(
